### DISCUSSION TITLE
Detecting if expression is in expression tree from analyzer

### DISCUSSION THREAD

**Discussion by kimsey0** [dotnet/roslyn#51446] [Q&A] [answered]
I'm building an analyzer that suggests changing `x == null` to `x is null`. Since pattern-matching isn't supported in expression trees, I try to filter those out via something like:
```csharp
var expression = context.Node.FirstAncestorOrSelf<ExpressionSyntax>(
    e => e is LambdaExpressionSyntax || e is QueryExpressionSyntax);
if (expression is not null)
{
    var expressionType = context.SemanticModel.GetTypeInfo(expression).ConvertedType?.OriginalDefinition;
    if (expressionType?.ToString().StartsWith("System.Linq.Expressions.Expression<") == true)
    {
        return;
    }
}
```
This works for lambda expressions, but not for query syntax expressions, where `GetTypeInfo` returns an `ErrorTypeSymbol` for a sample program like:

```csharp
using System.Linq;
var queryable = Enumerable.Empty<string?>().AsQueryable();
var query = from q in queryable where q == null select q;
```

Is there any reason Roslyn can't figure out the type of query expressions, or an alternative way to detect when they're converted to expression trees?

**Comment by jmarolf** [comment, marked as answer]
you should not be getting an error symbol in this case. Instead of doing a string comparison you should get the `System.Linq.Expressions.Expression` types from the compilation and then do symbol comparisons to ensure they are the same.

**Reply by kimsey0** [reply]
Thank you. With help from https://discord.gg/csharp, I figured out that I was missing [some](https://stackoverflow.com/a/47196516/2432538) metadata references, including the one for `System.Linq.Queryable.dll`, and had overlooked the diagnostics showing that. That was also why I was unable to get the `System.Linq.Expressions.Expression`1` type from the compilation and resorted to a crude string-based comparison.

**Comment by kimsey0** [comment]
For anyone else passing by this, I also ended up switching from a syntax analyzer to an operation analyzer, which allowed me to copy the logic from this internal utility: https://github.com/dotnet/roslyn-analyzers/blob/7bd2c98e798042e59121961640e124f76b03fdb9/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L518-L521 (Quite a few of the utilities there could probably do with being made public.)

The end result was something close to:
```csharp
var binaryOperation = (IBinaryOperation)context.Operation;

var expressionTreeType = context.Compilation.GetTypeByMetadataName("System.Linq.Expressions.Expression`1");
var lambdaAndLocalFunctionKinds = ImmutableArray.Create(OperationKind.AnonymousFunction, OperationKind.LocalFunction);
IOperation ancestor = binaryOperation;
do
{
    ancestor = ancestor.Parent;
} while (ancestor != null && !lambdaAndLocalFunctionKinds.Contains(ancestor.Kind));

if (SymbolEqualityComparer.Default.Equals(ancestor?.Parent?.Type?.OriginalDefinition, expressionTreeType))
{
    return;
}
```

**Reply by jnm2** [reply]
@kimsey0 I've recently done the same thing and realized that you can't stop after finding the first IAnonymousFunctionOperation ancestor, because it might not be for an expression tree lambda, but it might still be contained inside an expression tree lambda. For example, from my test:

```cs
using System.Collections.Generic;
using System.Linq;

class C
{
    public IEnumerable<int> RelatedCollection { get; }

    IQueryable<List<string>> M(IQueryable<C> firstQueryable, IQueryable<string> secondQueryable)
    {
        return firstQueryable.Select(firstElement =>
            firstElement.RelatedCollection.SelectMany(c => secondQueryable).ToList());
    }
}
```

and:

```cs
using System.Collections.Generic;
using System.Linq;

class C
{
    public IEnumerable<int> RelatedCollection { get; }

    IQueryable<List<string>> M(IQueryable<C> firstQueryable, IQueryable<string> secondQueryable)
    {
        return
            from firstElement in firstQueryable
            select firstElement.RelatedCollection.SelectMany(c => secondQueryable).ToList();
    }
}
```

This is my current way of handling it (`.Ancestors` and `.HasFullName` are extension methods I created):

```cs
using System.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Operations;

internal static class OperationFacts
{
    public static bool IsInsideExpressionTree(IOperation operation)
    {
        return operation.Ancestors()
            .Select(ancestor => (ancestor as IAnonymousFunctionOperation)?.Parent as IConversionOperation)
            .FirstOrDefault(functionConversion => functionConversion is not null) is
            {
                Type: { OriginalDefinition: { } anonymousFunctionType },
            }
            && anonymousFunctionType.HasFullName("System", "Linq", "Expressions", nameof(System.Linq.Expressions.Expression));
    }
}
```

**Reply by jnm2** [reply]
I also considered registering for operation blocks instead of operations, and using a OperationWalker subclass that skipped over anonymous function operations from the top level. After discussing with @CyrusNajmabadi, he suggested sticking with registering for the four operation kinds and doing this ancestor check last, on the principle that you should do the checks first that filter out the most cases.
